pkg/adapterManager/parallelManager.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adapterManager
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"context"
+
+	"istio.io/mixer/pkg/aspect"
+	"istio.io/mixer/pkg/attribute"
+	"istio.io/mixer/pkg/config"
+)
+
+// ParallelManager is an istio.io/mixer/pkg/api.Executor which wraps a Manager. Its BulkExecute method calls the
+// wrapped Manager's Execute in parallel across a pool of go routines. The pool of worker go routines is shared by all
+// request go routines, i.e. it's global.
+type ParallelManager struct {
+	*Manager
+
+	work chan<- task     // Used to hand tasks to the workers in the pool
+	quit chan<- struct{} // Used to shutdown the workers in the pool
+	wg   *sync.WaitGroup // Used to block shutdown until all workers complete
+}
+
+// NewParallelManager a new Manager who's BulkExecute method calls manager.Execute on each config in parallel.
+// size is the number of worker go routines in the worker pool the Manager schedules on to. This pool of workers is global:
+// the size of the pool should be roughly (max outstanding requests allowed)*(number of configs executed per request).
+func NewParallelManager(manager *Manager, size int) *ParallelManager {
+	work := make(chan task, size)
+	quit := make(chan struct{})
+	p := &ParallelManager{
+		Manager: manager,
+		work:    work,
+		quit:    quit,
+		wg:      &sync.WaitGroup{},
+	}
+
+	p.wg.Add(size)
+	for i := 0; i < size; i++ {
+		go p.worker(work, quit, p.wg)
+	}
+
+	return p
+}
+
+// BulkExecute takes a set of configurations and Executes all of them in parallel.
+func (p *ParallelManager) BulkExecute(ctx context.Context, cfgs []*config.Combined, attrs attribute.Bag) ([]*aspect.Output, error) {
+	numCfgs := len(cfgs)
+	// TODO: since we don't currently implement using different adapter sets per request,
+	// len(cfgs) is constant of the life of the configuration for all requests. We should probably use a
+	// pool for both the array of aspect outputs here and the result channels from requestGroup.
+	results := make([]*aspect.Output, numCfgs)
+
+	r, enqueue := p.requestGroup(attrs, numCfgs)
```

Comment:
could the requestGroup just take `[]cfg` as an argument?
since euque is an async operation, the main cancel point here is on line `81`.